diff --git a/src/video/SDL_egl.c b/src/video/SDL_egl.c
index c90380566..93d2705c2 100644
--- a/src/video/SDL_egl.c
+++ b/src/video/SDL_egl.c
@@ -40,10 +40,10 @@
 
 #if SDL_VIDEO_DRIVER_RPI
 /* Raspbian places the OpenGL ES/EGL binaries in a non standard path */
-#define DEFAULT_EGL "/opt/vc/lib/libEGL.so"
-#define DEFAULT_OGL_ES2 "/opt/vc/lib/libGLESv2.so"
-#define DEFAULT_OGL_ES_PVR "/opt/vc/lib/libGLES_CM.so"
-#define DEFAULT_OGL_ES "/opt/vc/lib/libGLESv1_CM.so"
+#define DEFAULT_EGL "/usr/lib/libEGL.so"
+#define DEFAULT_OGL_ES2 "/usr/lib/libGLESv2.so"
+#define DEFAULT_OGL_ES_PVR "/usr/lib/libGLES_CM.so"
+#define DEFAULT_OGL_ES "/usr/lib/libGLESv1_CM.so"
 
 #elif SDL_VIDEO_DRIVER_ANDROID || SDL_VIDEO_DRIVER_VIVANTE
 /* Android */
